test/test_webrtc_connection_mixin.py

```diff
@@ -0,0 +1,272 @@
+    async def test_concurrency_limit_reached_two_peers(self, test_client_and_stream):
+        test_client, stream = test_client_and_stream
+        pc1, channel = await self.setup_peer_connection(video=True)
+        pc2, channel = await self.setup_peer_connection(video=True)
+        await self.send_offer(pc1, test_client)
+        status, metadata = await self.send_offer(pc2, test_client, return_status_and_metadata=True)
+
+        assert status == 'failed'
+        assert metadata == {'error': 'concurrency_limit_reached', 'limit': 1}
+
+        await self.close_peer_connection(pc1)
+        await self.close_peer_connection(pc2)
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize("test_client_and_stream", [{"concurrency_limit": 2}], indirect=True)
+    @pytest.mark.xfail(reason="It is the same webrtc_id so pc1 and pc2 are seen as ONE connection")
+    async def test_concurrency_limit_reached_three_peers_same_id(self, test_client_and_stream):
+        test_client, stream = test_client_and_stream
+        pc1, channel = await self.setup_peer_connection(video=True)
+        pc2, channel = await self.setup_peer_connection(video=True)
+        pc3, channel = await self.setup_peer_connection(video=True)
+        await self.send_offer(pc1, test_client)
+        await self.send_offer(pc2, test_client)
+        status, metadata = await self.send_offer(pc3, test_client, return_status_and_metadata=True)
+
+        assert status == 'failed'
+        assert metadata == {'error': 'concurrency_limit_reached', 'limit': 1}
+
+        await self.close_peer_connection(pc1)
+        await self.close_peer_connection(pc2)
+        await self.close_peer_connection(pc3)
+
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize("test_client_and_stream", [{"concurrency_limit": 2}], indirect=True)
+    async def test_concurrency_limit_reached_three_peers(self, test_client_and_stream):
+        test_client, stream = test_client_and_stream
+        pc1, channel = await self.setup_peer_connection(video=True)
+        pc2, channel = await self.setup_peer_connection(video=True)
+        pc3, channel = await self.setup_peer_connection(video=True)
+        await self.send_offer(pc1, test_client, webrtc_id='foo')
+        await self.send_offer(pc2, test_client, webrtc_id='bar')
+        status, metadata = await self.send_offer(pc3, test_client, webrtc_id='baz', return_status_and_metadata=True)
+
+        assert status == 'failed'
+        assert metadata == {'error': 'concurrency_limit_reached', 'limit': 2}
+
+        await self.close_peer_connection(pc1)
+        await self.close_peer_connection(pc2)
+        await self.close_peer_connection(pc3)
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize("test_client_and_stream", [{"concurrency_limit": 1}], indirect=True)
+    @pytest.mark.xfail(reason="No audio or video tracks in peer connections added only datachannels")
```

Comment:
This is an interesting case. As long as a datachannel is added to the peer request it is a valid offer; however, they will not be added to ``stream.connections``.

Personally, I believe the best case would be to catch that no valid media stream(s) (depending on modality) is present, and close the connection, i.e., if you have ``modality=audio`` you need to supply an audio stream within a time limit, likewise, if you have ``modality="audio-video"`` both must be present.

Reply:
I think we should return the error if a peer connection has been received with a previous id. regardless of whether the tracks have been sent or the modality. Will push as well.